[Mathan Chidambaranathan] iP

ncmathan opened the pull request:
# Mathan
## Feature's of Mathan - the chatbot to organise tasks
1.  Add Todo items
    - Keeps track of task description
3. Add Events
    - Keeps track of event name
    - Event start time
    - Event end time
5. Add Deadlines
    - Keeps track of task description
    - task deadline date
 
Here's my run method in Duke.java:
```java
public void run(){
        ui.showWelcome();
        boolean isExit = false;
        while (!isExit) {
            String cmd = ui.readCommand();
            parser.parse(cmd, tasks, storage, ui);
            isExit = parser.isExit();
        }
        ui.showExit();
    }
```
You can refer to the `main` method to see how the `Duke` object is instantiated.

Here is what's done so far:
(updated as of 4 Sep 2023)
- [X] Level 1-9 :partying_face:
- [ ] Level 10
- [ ] A-Checkstyle
- [ ] A-Varargs
- [ ] ~~Integrate ChatGPT to automatically organise the tasks~~



> Give someone a program; you frustrate them for a day; teach them how to program, and you frustrate them for a lifetime – David Leinweber.

Here is the [reference](https://devtechnosys.com/insights/inspiring-programming-quotes-for-developer/) to the quote above

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

ncmathan committed:
no message

ncmathan committed:
Echo

ncmathan committed:
Add, List

ncmathan committed:
Mark as Done

ncmathan committed:
ToDo, Event, Deadline

ncmathan committed:
no message

ncmathan committed:
no message

ncmathan committed:
no message

ncmathan committed:
Save

ncmathan committed:
Save

ncmathan committed:
Revert "Save"

This reverts commit 1424bf707cba6489c6142199470368da194ad635.

ncmathan committed:
Merge branch 'branch-Level-7'

ncmathan committed:
Dates

ncmathan committed:
Revert "Dates"

This reverts commit 32d85e500887bd0ebe04c68b1257328d6ff6406c.

ncmathan committed:
Dates

ncmathan committed:
no message

ncmathan committed:
no message

ncmathan committed:
Merge branch 'branch-Level-8'

ncmathan committed:
extra OOP

ncmathan committed:
package

ncmathan committed:
gradle

ncmathan committed:
Add StorageTest.java and TaskListTest.java

Testing one method from each of the 2 classes.

ncmathan committed:
Add javadocs to all Java files

ncmathan committed:
Edit 7 files to follow coding standard

ncmathan committed:
Implement find method

ncmathan committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/Duke/Deadline.java
#	src/main/java/Duke/Duke.java
#	src/main/java/Duke/DukeException.java
#	src/main/java/Duke/Event.java
#	src/main/java/Duke/Parser.java
#	src/main/java/Duke/Storage.java
#	src/main/java/Duke/Task.java

ncmathan committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/Duke/Deadline.java
#	src/main/java/Duke/Duke.java
#	src/main/java/Duke/DukeException.java
#	src/main/java/Duke/Event.java
#	src/main/java/Duke/Parser.java
#	src/main/java/Duke/Storage.java
#	src/main/java/Duke/Task.java

ncmathan committed:
Merge branch 'branch-A-JavaDoc'

ncmathan committed:
Fix minor bugs

ncmathan committed:
implement GUI

zhonghan721 commented on src/main/java/Duke/Duke.java:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    public Duke(){
+        ui=new Ui();
+        storage=new Storage("file.txt");
+        tasks= new TaskList(storage.load());
+        parser=new Parser();
```
Perhaps some whitespace between the assignment operator might be better?

zhonghan721 commented on src/main/java/Duke/Event.java:
```diff
@@ -0,0 +1,29 @@
+        this.to = LocalDate.parse(to,DateTimeFormatter.ofPattern("MMM dd yyyy"));
+    }
+    public String toString(){
+        return "[E]"+super.toString() +" (from: "+from.format(DateTimeFormatter.ofPattern("MMM dd yyyy"))+" to: "+to.format(DateTimeFormatter.ofPattern("MMM dd yyyy"))+")";
```
A line break in this return statement can perhaps improve readability.

zhonghan721 commented on src/main/java/Duke/Storage.java:
```diff
@@ -0,0 +1,145 @@
+            while ((st = br.readLine()) != null){
+                String[] out=st.split(",");
+                switch (out[0]){
+                    case "T":
```
The indentation here seems to be following the default IDE configuration.

zhonghan721 commented on src/main/java/Duke/Storage.java:
```diff
@@ -0,0 +1,145 @@
+                if(count!=num)
+                    out+=line+"\n";
+                else{
+                    if(val)
+                        out+=line.substring(0,2)+"1"+line.substring(3)+"\n";
+                    else
+                        out+=line.substring(0,2)+"0"+line.substring(3)+"\n";
```
Perhaps some curly brackets for conditionals might improve readability and debugging?

zhonghan721 commented on src/main/java/Duke/TaskList.java:
```diff
@@ -0,0 +1,206 @@
+            list.get(index-1).setDone();
+            storage.updateFile(index,true);
+            ui.print("Nice! I've marked this task as done:");
+        }
+        else{
```
Egyptian style brackets might make the code more tidy here.

zhonghan721 commented on src/main/java/Duke/Duke.java:
```diff
@@ -0,0 +1,99 @@
+    /**
+     * Contains the logic essential to the running of Duke.
+     */
+    public void run(){
+        ui.showWelcome();
+        boolean isExit = false;
+        while (!isExit) {
+            String cmd = ui.readCommand();
+            parser.parse(cmd, tasks, storage, ui);
+            isExit = parser.isExit();
+        }
+        ui.showExit();
+    }
```
I like how clean this bit of code is 👍

leezhengjing commented on src/main/java/Duke/MainWindow.java:
```diff
@@ -0,0 +1,64 @@
+import javafx.scene.image.Image;
+import javafx.scene.layout.AnchorPane;
+import javafx.scene.layout.VBox;
+/**
```
Maybe could add a new line between the imports and the class header comment

leezhengjing commented on src/main/java/Duke/Parser.java:
```diff
@@ -0,0 +1,58 @@
+                list.find(str,ui);
+            } else {
+                ui.handleError(new DukeException("\n____________________________________________________________\n" +
+                        "☹ OOPS!!! I'm sorry, but I don't know what that means :-(\n" +
```
Maybe could make better use of the Ui class and abstract all the text output to specific methods of the Ui class, instead of having a singular handleError method

leezhengjing commented on src/main/java/Duke/TaskList.java:
```diff
@@ -0,0 +1,206 @@
+        if(mark){
+            list.get(index-1).setDone();
+            storage.updateFile(index,true);
+            ui.print("Nice! I've marked this task as done:");
```
Like the previous comment, I think the Ui class could be better abstracted. i.e. ui.printTaskDone(); instead of having to write the text here, as it would be a hassle to find in the future to update your text outputs if you decide to change the way your Ui responds.

leezhengjing commented on src/main/java/Duke/TaskList.java:
```diff
@@ -0,0 +1,206 @@
+        ui.showLine();
+    }
+
+    public void find (String str, Ui ui) {
```
I think theres an unintended space in the find () method

leezhengjing commented on src/main/java/Duke/Ui.java:
```diff
@@ -0,0 +1,87 @@
+     * Default constructor.
+     */
+    public Ui(){
+        buffer = "";
+        scanner=new Scanner(System.in);
```
Inconsistent space between assignment operator.

leezhengjing reviewed (commented):
Overall looks good just some slight styling errors and would suggest to improve the Ui class functionality.

elhy1999 commented on src/main/java/Duke/Parser.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    /**
+     * static method which takes in instances of other essential classes.
```
Should this method not be classified as a "static method" since the method doesn't have a static keyword? 🤔

elhy1999 commented on src/main/java/Duke/Task.java:
```diff
@@ -0,0 +1,40 @@
+        return isDone;
+    }
+
+    public boolean contains (String val) {
```
Perhaps this method name could be changed to a verb instead, like getContains or something of the sort? 😃

elhy1999 commented on src/main/java/Duke/TaskList.java:
```diff
@@ -0,0 +1,206 @@
+    public void todo(String str, Ui ui, Storage storage){
+        str = str.substring(5);
+        if (str.isEmpty())
+            ui.handleError(new DukeException("\n____________________________________________________________\n" +
```
You seem to be using these horizontal lines a couple of times throughout your code. Perhaps creating a global variable for this horizontal line may help? 😄

elhy1999 commented on src/main/java/Duke/Storage.java:
```diff
@@ -0,0 +1,145 @@
+import java.util.ArrayList;
+import java.util.Scanner;
+
+/**
```
I like how you have documented almost every single class you created 👍

ncmathan committed:
Add various assert statements

Ensured the parameters passed to initialising the various tasks are not
empty. Ensured the date format is correct for Deadline and Event.
Ensured dialogContainer only has one child when initialised.

Removed a bug that causes errors to not show up immediately in the Duke
chatbot and instead wait for valid input from user before showing all
errors simultaneously.

ncmathan committed:
Improve Code Quality

Removed unused variables and methods. Removed indentations from
switch statements and added default statement. Wrapped lines of code
that were too long. Made happy path more prominent.

Fixed a bug where DukeException was not printed to chatbot immediately.

ncmathan committed:
Merge pull request #1 from ncmathan/branch-A-Assertions

Add various assert statements

ncmathan committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/Duke/DialogBox.java
#	src/main/java/Duke/Duke.java

ncmathan committed:
Merge branch 'master' into branch-A-CodeQuality

ncmathan committed:
Merge pull request #2 from ncmathan/branch-A-CodeQuality

Improve Code Quality

ncmathan committed:
Merge branch 'branch-A-CodeQuality' of https://github.com/ncmathan/ip into branch-A-CodeQuality

ncmathan committed:
add sort functionality

Added 2 more commands which can be parsed: "sort deadline" and "sort
event". It lists the respective tasks in chronological order from
earliest to latest.

ncmathan committed:
Merge branch 'branch-A-CodeQuality'

ncmathan committed:
Update Readme docs. Minor bug fixed

ncmathan committed:
include duke.jar in src folder

ncmathan committed:
modified readme

ncmathan committed:
Removed README from root folder

ncmathan committed:
update docs/README.md

ncmathan committed:
Update README.md

ncmathan committed:
add more Javadocs

ncmathan committed:
Merge branch 'master' of https://github.com/ncmathan/ip
